policybot/pkg/testflakes/testflakes.go

```diff
@@ -0,0 +1,358 @@
+}
+
+/*
+ * Function to initionalize new PrFlakeTest object.
+ * Use `prFlakeyTest, err := NewPrFlakeTest()` and `testFlakes := prFlakeyTest.checkTestFlakesForPr(<pr_number>)`
+ * to get test flakes information for a given pr.
+ */
+func NewPrFlakeTest() (*PrFlakeTest, error) {
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &PrFlakeTest{
+		client: client,
+		ctx:    ctx,
+	}, nil
+}
+
+/*
+ * Contains function check if a string exists in a given slice of strings.
+ */
+func contains(slic []string, ele string) bool {
+	for _, e := range slic {
+		if strings.Compare(e, ele) == 0 {
+			return true
+		}
+	}
+	return false
+}
+
+/*
+ * GetTest function get all directories under the given pr in istio-prow/pr-logs/pull/istio-istio/PRNUMBER for each test suite name.
+ * Client: client used to get buckets and objects.
+ * PrNum: the PR number inputted.
+ * Return []Tests return a slice of Tests objects.
+ */
+func (prFlakeTest PrFlakeTest) getTests(prNumInt int64) ([]Tests, error) {
+	prNum := strconv.FormatInt(prNumInt, 10)
+	ctx := prFlakeTest.ctx
+	client := prFlakeTest.client
+	bucket := client.Bucket("istio-prow")
+	query := &storage.Query{Prefix: "pr-logs/pull/istio_istio/" + prNum}
+	it := bucket.Objects(ctx, query)
+	var testSlice []Tests
+	for {
+		attrs, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		name := attrs.Name
+
+		nameSlice := strings.Split(name, "/")
+		prNum = nameSlice[3]
+		pullNum := nameSlice[5]
+		testName := nameSlice[len(nameSlice)-3]
+		fileName := nameSlice[len(nameSlice)-1] // C
+		var newString = "pr-logs/pull/istio_istio/" + prNum + "/" + testName + "/" + pullNum
+		if strings.Compare(fileName, "started.json") == 0 || strings.Compare(fileName, "clone-records.json") == 0 || strings.Compare(fileName, "finished.json") == 0 {
+			var contain = false
+			for ind, ele := range testSlice {
+				if strings.Compare(ele.Name, testName) == 0 {
+					prs := ele.Prs
+
+					if !contains(prs, newString) {
+						prs = append(prs, newString)
+						ele.Prs = prs
+						testSlice[ind] = ele
+					}
+
+					contain = true
+				}
+
+			}
+			if !contain {
+				t := Tests{
+					Name: testName,
+				}
+				newSlice := []string{newString}
+				t.Prs = newSlice
+				testSlice = append(testSlice, t)
+			}
+		}
+	}
+	return testSlice, nil
+}
+
+/*
+ * GetShaAndPassStatus function return the status of test passing, clone failure, sha number, base sha for each test run under each test suite for the given pr.
+ * Client: client used to get buckets and objects from google cloud storage.
+ * TestSlice: a slice of Tests objects containing all tests and the path to folder for each test run for the test under such pr.
+ * Return a map of test suite name -- pr number -- run number -- ForEachRun objects.
+ */
+func (prFlakeTest PrFlakeTest) getShaAndPassStatus(testSlice []Tests) ([]*store.TestFlake, error) {
```

Comment:
Could you break this function into smaller pieces? It's a bit large as is.

Also, it's generally not a great idea to use 'defer' within a loop, since the deferred close only happens at function end, and not at loop end. If you split this up into subfunctions, then the defer can happen must closer to the optimal point.